### PR TITLE
add Signer to api

### PR DESCRIPTION
**Pull request by ianhe8x**
see #660

### CODE REVIEW

File: packages/api/test/util/SimpleSigner.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+
+export default class SimpleSigner{
```
Comment (jacogr) [inline packages/api/test/util/SimpleSigner.js]: I would suggest renaming this one to `SingleAccountSigner` or just change it slightly to take in a keyring instance and then doing `keyring.getPair(options.from)` (Which is what I assumed the `SimpleSigner` to be, but maybe that is rather a `KeyringSigner` - the above suggested naming I would probably apply here anyway, just to be explicit)

File: packages/api/src/SubmittableExtrinsic.ts [hunk 2]
```diff
@@ -163,25 +163,40 @@ export default class SubmittableExtrinsic<CodecResult, SubscriptionResult> exten
...
+      ).pipe(
+        first(),
+        mergeMap(async (nonce) => {
+          if(isKeyringPair){
+            this.sign(account as KeyringPair, { ...options, nonce });
+          }else{
+            if(!this._api.signer){
+              throw new Error('no signer exists');
+            }
+            await this._api.signer.sign(this, {
```
[the remark this answers is not in the document]
Comment (ianhe8x) [reply packages/api/src/SubmittableExtrinsic.ts]: I agree that making signer interface `sign(extrinsic, address, options)` keeps kind of consistency with signAndSend. And I can try extracting the logic of expanding options into a build function and let's see if it improves.

File: packages/api/test/util/SimpleSigner.js (hunk 1 above)
Comment (ianhe8x) [reply packages/api/test/util/SimpleSigner.js]: Just a reminder that this class is only supposed to be used in test. I don't think we need to ship a implementation of Signer with api. (From your comments I feel you think this is a exported class.)

File: packages/api/src/Base.ts [hunk 3]
```diff
@@ -101,6 +101,7 @@ export default abstract class ApiBase<CodecResult, SubscriptionResult> implement
     this._rpcRx = new RpcRx(options.provider);
     this._rpc = this.decorateRpc(this._rpcRx, this.onCall);
     this._rx.rpc = this.decorateRpc(this._rpcRx, rxOnCall);
+    this._rx.signer = options.signer;
```
Comment (amaurymartiny) [inline packages/api/src/Base.ts]: imo signer has nothing to do with rx, so it shouldn't be in rx. I would propose `this._signer = options.signer` here

Comment (ianhe8x) [reply packages/api/src/Base.ts]: it is more like `_rx` is not a proper name, because it's the api object that is passed to SubmittableExtrinsic.

**Comment by ianhe8x** [comment]
@jacogr when can this pr get merged?

**Comment by jacogr** [comment]
Sorry, they have been heaping up a bit, all of them. (I have been a bottleneck this week)

**Pull request merged**